### PR TITLE
Created a single-threaded task stream

### PR DESCRIPTION
**Pull request by ryanpbrewster**
@leogrim @andrewconner

### CODE REVIEW

File: kifi-backend/modules/common/test/com/keepit/common/concurrent/TaskStreamTest.scala [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+class TaskStreamTest extends Specification with Logging {
+
+  implicit val execCtx = ExecutionContext.fj
+
+  "TaskStream" should {
+    "run futures lazily" in {
+      "simple squaring" in {
+        val processed = mutable.ArrayBuffer.empty[Int]
+        def slowFut(x: Int) = Future {
+          Thread.sleep(10 * x)
```
Comment (andrewconner) [inline kifi-backend/modules/common/test/com/keepit/common/concurrent/TaskStreamTest.scala:20]: Any way to avoid this?

**Comment by andrewconner** [comment]
Mind going over the motivation a bit?

**Comment by ryanpbrewster** [comment]
Fiddling with a few ways to unify the `FutureHelpers` API a bit; trying to make it look more like the standard Scala collections API.

File: kifi-backend/modules/common/test/com/keepit/common/concurrent/TaskStreamTest.scala (hunk 1 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/common/test/com/keepit/common/concurrent/TaskStreamTest.scala:20]: Sure, it was mostly there to reassure me that the tasks were executed sequentially (as opposed to something like `Future.sequence`)

**Comment by ryanpbrewster** [comment]
Example use case: suppose you are working with Slack's channel history API and you want to count the number of links since a given message timestamp.

```
val numLinksFut = FutureHelpers.foldLeftUntil(Stream.continually(()))((initialTimestamp, 0)) {
    case ((workingTimestamp, numLinksSoFar), _) =>
        slackClient.channelHistory(workingTimestamp, pageSize).map { msgs =>
            ((msgs.map(_.timestamp).maxOpt, numLinksSoFar + countLinksInMessages(msgs), msgs.isEmpty)
        }
}
```

vs

```
val numLinksFut = SyncList.accumulate(initialTimestamp) { workingTimestamp =>
    slackClient.channelHistory(workingTimestamp, pageSize).imap { msgs =>
        (msgs, msgs.map(_.timestamp).maxOpt)
    }
}.takeWhile(_.nonEmpty).map(countLinksInMessages).foldLeft(0)(_ + _).run
```

Haven't actually tried compiling those, but that's the dream

**Comment by andrewconner** [comment]
I see, cool. This seems like the motivation behind scalaz-streams and akka-streams, you may like them (I've played with both, I like akka's better)

**Comment by ryanpbrewster** [comment]
Yeah, I think I may just be poorly implementing those. I'll take a look.

PSA: This isn't a super serious PR, just wanted to mess around with it and get some feedback.

File: kifi-backend/modules/common/app/com/keepit/common/concurrent/SyncList.scala [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+import java.util.NoSuchElementException
+
+import scala.concurrent.{ Future, ExecutionContext => ScalaExecutionContext }
+import com.keepit.common.core.futureExtensionOps
+
+class Task[T](f: () => Future[T]) {
+  lazy val run: Future[T] = f()
+  def map[S](g: T => S): Task[S] = new Task(() => f().imap(g))
+  def flatMap[S](g: T => Future[S])(implicit exc: ScalaExecutionContext): Task[S] = new Task(() => f().flatMap(g))
+}
```
Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/common/concurrent/SyncList.scala:12]: I feel like using `map` / `flatMap` isn't appropriate, will lead to unintended memory leaks

**Pull request closed without merging**